### PR TITLE
[integration] develop/admin documentation about OAuth2 authorization

### PR DESCRIPTION
**Pull request by TaykYoku**
Add documentation about OAuth2 authorization components. Wait #5540, #5398.

BEGINRELEASENOTES

*docs
NEW: OAuth2 authorization components documentation

ENDRELEASENOTES

### CODE REVIEW

File: docs/source/DeveloperGuide/OAuth2Authorization/index.rst [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+OAuth 2.0 roles
+===============
+
+An OAuth 2.0 flow has the following `roles <https://datatracker.ietf.org/doc/html/rfc6749#section-1.1>`_:
+
+ - **Resource Owner** - Entity that can grant access to a protected resource. In the context of DIRAC, these are DIRAC users.
+ - **Resource Server** - Server hosting the protected resources. In the context of DIRAC, this is DIRAC backend components like a DIRAC services.
+ - **Client** - Application requesting access to a protected resource on behalf of the *Resource Owner*. In the context of DIRAC, these are DIRAC client installations.
+ - **Authorization Server** - Server that authenticates the *Resource Owner* and issues access tokens after getting proper authorization. In the context of DIRAC, this is DIRAC Authorization Server.
+
```
Comment (fstagni) [inline docs/source/DeveloperGuide/OAuth2Authorization/index.rst:37]: In this schema, where do DIRAC agents fit?

We have DIRAC agents that authenticate themselves using the server certificate of the machine where they are running, while some use a proxy of a user (from the ProxyManager) to authenticate to external services.

File: docs/source/DeveloperGuide/OAuth2Authorization/index.rst [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+ - :ref:`dirac-login`
+ - :ref:`dirac-logout`
+
+Also added the ability to authorize without a certificate while configuring the DIRAC client with the :ref:`dirac-configure` command and a special ``--login`` flag.
+
+
+Authorization API
+=================
+
+With a new system component - :ref: `APIs <apis>`, was created Authorization API for =Framework= system (see :py:class:`~DIRAC.FrameworkSystem.API.AuthHandler`) which provides the necessary endpoints for interaction with DIRAC AS.
```
Comment (fstagni) [inline docs/source/DeveloperGuide/OAuth2Authorization/index.rst:130]: `=Framework=`? I think you meant `*Framework*`. (md vs rst :) )

**Review by fstagni — COMMENTED** [review]
Thank you! It's shaping good.

File: docs/source/DeveloperGuide/OAuth2Authorization/index.rst (hunk 2 above)
Comment (TaykYoku) [reply docs/source/DeveloperGuide/OAuth2Authorization/index.rst:130]: yes, thanks. Ill fix it

File: docs/source/DeveloperGuide/OAuth2Authorization/index.rst (hunk 1 above)
Comment (TaykYoku) [reply docs/source/DeveloperGuide/OAuth2Authorization/index.rst:37]: When an **Agent** (or **Service**) makes a request to another **Service**, in the _OAuth2_ paradigm, the **Agent** acts as a **Client**, a **Service** accessed by the **Agent** - a **resource server** or a **protected resource**.

If the **Agent** uses a _host certificate_, this is a type of access similar to accessing with _client credentials_ (when _client_id_ and _client_secret_ are passed in requests) and is a request for a **protected resource** on behalf of the **client**. If an **agent** uses a _proxy_ (or user _access token_), it is a request for a **protected resource** on behalf of the **resource owner**.

It looks like I need to add one more diagram showing the interaction between DIRAC components (without direct user involvement).

Comment (fstagni) [reply docs/source/DeveloperGuide/OAuth2Authorization/index.rst:37]: > It looks like I need to add one more diagram showing the interaction between DIRAC components (without direct user involvement).

Thanks! that would help!

Comment (TaykYoku) [reply docs/source/DeveloperGuide/OAuth2Authorization/index.rst:49]: added

File: docs/source/DeveloperGuide/OAuth2Authorization/index.rst [hunk 3]
```diff
@@ -0,0 +1,281 @@
...
+
+OAuth 2.0 Device flow:
+  1) ``dirac-login`` initializes `OAuth 2.0 Device flow` by passing DIRAC client ID to DIRAC AS.
+  #) DIRAC AS responds with a ``device_code``, ``user_code``, ``verification_uri``, ``verification_uri_complete``, ``expires_in`` (lifetime in seconds for device_code and user_code), and polling ``interval``.
+  #) The command asks the user to log in using a device that has a browser(e.g.: their computer, smartphone) or if the device running ``dirac-login`` has a browser installed, a new tab with the received URL will open automatically.
+
+    a) The command begins polling DIRAC AS for an access token sending requests to token endpoint until either the user completes the browser flow path or the user code expires.
+
+OAuth 2.0 Authorization Code flow:
+  4) After receiving this request from the browser, DIRAC AS will initialize ``OAuth 2.0 Authorization Code`` flow with choosed IdP. If several IdPs are registered in DIRAC and it is not clear from the requested group which one to choose, DIRAC AS will ask the user to choose one.
```
Comment (fstagni) [inline docs/source/DeveloperGuide/OAuth2Authorization/index.rst:210]: shouldn't this start with 1)?

File: docs/source/DeveloperGuide/OAuth2Authorization/index.rst [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+
+DIRAC CLI
+=========
+
+Using ``dirac-logout``:
+
+.. image:: /_static/Systems/FS/OAuth2/revokeToken.png
+   :alt: DIRAC logout flow.
+
+If it is a long session, ie with a refresh token, which allows you to update the access token and thus continue the working session, then to end the session it is necessary to revoke refresh token:
```
Comment (fstagni) [inline docs/source/DeveloperGuide/OAuth2Authorization/index.rst:264]:
```suggestion
If it is a long session, i.e. with a refresh token, which allows you to update the access token and thus continue the working session, then to end the session it is necessary to revoke the refresh token:
```

File: docs/source/DeveloperGuide/OAuth2Authorization/index.rst [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+  1) :ref:`dirac-logout` sends a revoke request to DIRAC AS.
+
+    a) DIRAC AS decrypts the refresh token and reads to whom it belongs.
+    #) DIRAC AS makes a revoke request to the appropriate IdP.
+    #) DIRAC AS remove record about this refresh token in ``AuthDB`` database.
```
Comment (fstagni) [inline docs/source/DeveloperGuide/OAuth2Authorization/index.rst:269]:
```suggestion
    #) DIRAC AS removes the record about this refresh token from the ``AuthDB`` database.
```

File: docs/source/DeveloperGuide/OAuth2Authorization/index.rst [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+   - public clients
+
+  DIRAC AS takes both into account and already has a default *public client* (see :py:class:`~DIRAC.FrameworkSystem.private.authorization.utils.Clients`) configured to authorize DIRAC client installations via the device code authorization flow mentioned earlier.
+  The new `authorization client metadata <https://datatracker.ietf.org/doc/html/rfc7591#section-2>`_ can be described in the `/DIRAC/Security/Authorization/Clients` section in format::
+
+      CLIENT_NAME
+      {
+        client_id=MY_CLIENT_ID
+        client_secret=MY_CLIENT_SECRET
+        scope=supported scopes separated by a space
```
[the remark this answers is not in the document]
Comment (TaykYoku) [reply docs/source/DeveloperGuide/OAuth2Authorization/index.rst:116]: no, the value of this option is scope as it is.  that is, the list is not expected to be returned

**Comment by fstagni** [comment]
Could you squash the commits?

File: docs/source/AdministratorGuide/ServerInstallations/environment_variable_configuration.rst [hunk 7]
```diff
@@ -77,4 +77,4 @@ BEARER_TOKEN_FILE
   If the environment variable is set, then its value is interpreted as a filename. The content of the specified file is used as token string (https://doi.org/10.5281/zenodo.3937438).
 
 DIRAC_USE_ACCESS_TOKEN
-  If this environment is set to ``true`` or ``yes``, the concurrent.futures.ThreadPoolExecutor will be used (default=false)
+  If this environment is set to ``true``, then when trying to connect to the server, access tokens obtained during authorization will be used (default=false)
```
[the remark this answers is not in the document]
Comment (TaykYoku) [reply docs/source/AdministratorGuide/ServerInstallations/environment_variable_configuration.rst:80]: This means that the token will be used, no matter how the user received it. Just like a proxy.
Fixed to:
```
  If this environment is set to ``true``, then when trying to connect to the server, access tokens will be used (default=false)
```

File: docs/source/DeveloperGuide/OAuth2Authorization/index.rst (hunk 3 above)
Comment (TaykYoku) [reply docs/source/DeveloperGuide/OAuth2Authorization/index.rst:210]: I changed the structure a bit to look holistic in one context

**Comment by fstagni** [comment]
RTD is failing.

**Comment by TaykYoku** [comment]
because it needs #5540, #5398.

**Comment by fstagni** [comment]
There's a conflict now.

**Review by fstagni — APPROVED** [review]

**Pull request merged**